src/nvme/linux.c

```diff
@@ -543,6 +559,187 @@ int nvme_gen_dhchap_key(char *hostnqn, enum nvme_hmac_alg hmac,
+
+	err = -ENOKEY;
+	if (EVP_PKEY_derive_init(ctx) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_set_hkdf_md(ctx, md) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_set1_hkdf_key(ctx, generated_key, key_len) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, "tls13 ", 6) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, "HostNQN", 7) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, hostnqn, strlen(hostnqn)) <= 0)
+		goto out_free_retained_key;
+
+	if (EVP_PKEY_derive(ctx, retained_key, &retained_len) <= 0) {
+		fprintf(stderr, "EVP_KDF derive failed\n");
+		err = ENOKEY;
+	}
+	err = 0;
+out_free_retained_key:
+	if (err) {
+		free(retained_key);
+		retained_key = NULL;
+	}
+	EVP_PKEY_CTX_free(ctx);
+	return retained_key;
+}
+
+static long int derive_tls_key(const EVP_MD *md, key_serial_t keyring,
+			       const char *hostnqn, const char *subsysnqn,
+			       unsigned char *retained_key, int key_len)
+{
+	key_serial_t key = -ENOKEY;
+	const char *key_type = "psk";
+	EVP_PKEY_CTX *ctx;
+	char *psk_identity;
+	unsigned char *psk;
+	size_t psk_len;
+	long int err;
+
+	psk_identity = malloc(strlen(hostnqn) + strlen(subsysnqn) + 12);
+	if (!psk_identity)
+		return ENOMEM;
+
+	sprintf(psk_identity, "NVMeR%02d %s %s", md == EVP_sha256() ? 1 : 2,
+		hostnqn, subsysnqn);
+
+	psk = malloc(key_len);
+	if (!psk) {
+		err = ENOMEM;
+		goto out_free_identity;
+	}
+	psk_len = key_len;
+
+	ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL);
+	if (!ctx) {
+		err = ENOMEM;
+		goto out_free_psk;
+	}
+
+	err = -ENOKEY;
+	if (EVP_PKEY_derive_init(ctx) <= 0)
+		goto out_free_ctx;
+	if (EVP_PKEY_CTX_set_hkdf_md(ctx, md) <= 0)
+		goto out_free_ctx;
+	if (EVP_PKEY_CTX_set1_hkdf_key(ctx, retained_key, key_len) <= 0)
+		goto out_free_ctx;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, "tls13 ", 6) <= 0)
+		goto out_free_ctx;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, "nvme-tls-psk", 12) <= 0)
+		goto out_free_ctx;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, psk_identity,
+					strlen(psk_identity)) <= 0)
+		goto out_free_ctx;
+
+	if (EVP_PKEY_derive(ctx, psk, &psk_len) <= 0) {
+		fprintf(stderr, "EVP_KDF_derive failed\n");
+	}
+
+	key = keyctl_search(keyring, key_type, psk_identity, 0);
+	if (key >= 0) {
+		printf("updating %s key '%s'\n",
+		       key_type, psk_identity);
+		err = keyctl_update(key, psk, psk_len);
+		if (err) {
+			fprintf(stderr, "updating %s key '%s' failed\n",
+				key_type, psk_identity);
+			key = err;
+		}
+	} else {
+		printf("adding %s key '%s'\n", key_type, psk_identity);
+		key = add_key(key_type, psk_identity,
+			      psk, psk_len, keyring);
+		if (key < 0)
+			fprintf(stderr, "adding %s key '%s' failed, error %d\n",
+				key_type, psk_identity, errno);
+	}
+
+out_free_ctx:
+	EVP_PKEY_CTX_free(ctx);
+out_free_psk:
+	free(psk);
+out_free_identity:
+	free(psk_identity);
+
+	return key;
+}
+
+long nvme_insert_tls_key(const char *keyring, const char *hostnqn,
```

Comment:
`int` return type?

Reply:
Weird keyutils semantic. What I'm actually want is to return a key serial (which is defined as 'long'), but returning a type 'key_serial_t' would give us a hit on keyutils.h, and it won't be optional anymore.
But I'll check if I can make the call to the function optional, then it should be possible to return a key_serial_t here.